app/models/concerns/doi_generator.rb

```diff
@@ -0,0 +1,170 @@
+module DoiGenerator
+  extend ActiveSupport::Concern
+
+  attr_accessor :doi_message
+
+  def check_doi_presence
+    return self.doi_message unless can_get_doi?
+    if self.doi.present?
+      update_datacite_doi
+    else
+      create_datacite_doi
+    end
+    self.doi_message
+  end
+
+  private
+
+  def can_get_doi?
+    # Only generate if required metadata is there
+    unless self.id.present? && self.creator.present? && self.title.present?
+      self.doi_message = 'metadata'
+      return false
+    end
+    # Only generate for combined pages
+    if self.is_a?(Page) && self.page_number.present?
+      self.doi_message = 'page'
+      return false
+    end
+    true
+  end
+
+  def create_datacite_doi
+    response = DataciteRest.new.mint(
+      datacite_api_json
+    )
+    data = JSON[response.body]["data"]
+    self.update_attributes(doi: [data["id"]])
+    self.doi_message = 'generated'
```

Comment:
What is `doi_message` used for? Is it something that existed in the previous DOI lookup that needs to be retained?

Reply:
I think it is used for our purpose, not sure if it needs to be retained. I can see it being useful to check DOI states in case they are not behaving as expected.